src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Reflection;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.References
+{
+    /// <summary>
+    /// Base reference context provider which abstract CPS overrides
+    /// </summary>
+    public class BaseReferenceContextProvider : IVsReferenceManagerUserAsync
+    {
+        /// <summary>
+        /// Value used to override the CPS provider
+        /// </summary>
+        public const int OverrideCPSProvider = 1;
+
+        /// <summary>
+        /// Lazy instance of the next handler in the chain.
+        /// </summary>
+        private Lazy<Lazy<IVsReferenceManagerUserAsync, IVsReferenceManagerUserComponentMetadataView>> nextHandler;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BaseReferenceContextProvider"/> class.
+        /// </summary>
+        [ImportingConstructor]
+        public BaseReferenceContextProvider(ConfiguredProject configuredProject)
+        {
+            this.VsReferenceManagerUsers = new OrderPrecedenceImportCollection<IVsReferenceManagerUserAsync, IVsReferenceManagerUserComponentMetadataView>(projectCapabilityCheckProvider: configuredProject);
+            this.nextHandler = new Lazy<Lazy<IVsReferenceManagerUserAsync, IVsReferenceManagerUserComponentMetadataView>>(() =>
+            {
+                Type provider = this.GetType();
+                var order = provider.GetCustomAttribute<OrderAttribute>();
+                var user = provider.GetCustomAttribute<ExportIVsReferenceManagerUserAsyncAttribute>();
+                return this.VsReferenceManagerUsers.FirstOrDefault(
+                        export =>
+                            export.Metadata.OrderPrecedence < order.OrderPrecedence &&
+                            export.Metadata.ProviderContextIdentifier == user.ProviderContextIdentifier);
+            });
+        }
+
+        /// <summary>
+        /// Gets the next handler in the chain.
+        /// </summary>
+        private IVsReferenceManagerUserAsync NextHandler
+        {
+            get
+            {
+                return nextHandler?.Value?.Value;
+            }
+        }
+
+        /// <summary>
+        /// Gets the collection of reference provider contexts that can handle individual reference type operations.
+        /// </summary>
+        [ImportMany]
```

Comment:
Removed the import on the config (which was already being imported in the constructor anyway). This type (OrderPrecedenceImportCollection) needs to be done via a property or MEF complains. There's many uses of it in the codebase in the same manner.